### ISSUE TITLE
[Feature request] client-level constraints

### ISSUE THREAD

**Issue by Legogris** [hashicorp/nomad#8663]
There are many scenarios where one would like to ensure that a particular node only receives a certain job or jobs. As I see it, there is no way of restricting nodes to particular jobs or classes of jobs without introducing constraints in some way on jobs themselves (e.g. using `meta` or `node_class`). There are workarounds, for example restricting those jobs to use different drivers that are only used for those jobs, but that's more of a hacky workaround.

It'd be great if the `client` stanza had some mechanism to whitelist certain jobs or task groups - be it by a `constraint` stanza, requiring certain `meta` fields, or even just a whitelist of job or task group names.

**Comment by tgross** [comment]
Hi @Legogris!

One way I've done this is to have those clients in a different [`datacenter`](https://www.nomadproject.io/docs/configuration#datacenter) (but still joined to the same set of servers). Because `datacenter` is already a required field in the jobspec, it gives you a very coarse-grained control.

**Comment by Legogris** [comment]
@tgross Thanks for the suggestion, that's not a bad idea. There's no impact to having this differ from consul dc for nomad clusters on consul, right?

**Comment by tgross** [comment]
> There's no impact to having this differ from consul dc for nomad clusters on consul, right?

Right! A Consul `datacenter` is actually semantically closer to a Nomad `region`, so typically you'll end up with one Consul DC per Nomad region.

**Comment by pznamensky** [comment]
Nevertheless the "datacenter" workaround is not a good decision for all cases: https://github.com/hashicorp/nomad/issues/2299#issuecomment-560536149

**Comment by Legogris** [comment]
> @tgross Thanks for the suggestion, that's not a bad idea. There's no impact to having this differ from consul dc for nomad clusters on consul, right?

This method does not work for all at me in the end; Unless There's something else going on it seems to have an impact on dc-level for Consul service discovery.

Basically I have all Consul clients and servers in the same Consul datacenter, with a Nomad server running Traefik using Consul Catalog service discovery on a separate Nomad datacenter.

The Consul client on that node errors with:

```
 agent.client: RPC failed to server: method=Catalog.ListServices server=10.1.2.33:8300 error="rpc error making call: No path to datacenter"
agent.http: Request error: method=GET url=/v1/catalog/services?dc=xxx from=127.0.0.1:15801 error="rpc error making call: No path to datacenter"
```
